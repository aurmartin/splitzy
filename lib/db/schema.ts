import { type RunResult } from "better-sqlite3";
import { type ExtractTablesWithRelations } from "drizzle-orm";
import { type BetterSQLite3Database } from "drizzle-orm/better-sqlite3";
import { type ExpoSQLiteDatabase } from "drizzle-orm/expo-sqlite";
import {
  SQLiteColumnBuilderBase,
  SQLiteTransaction,
  sqliteTable,
  text,
} from "drizzle-orm/sqlite-core";
import { type SQLiteRunResult } from "expo-sqlite";

export const groupsTable = sqliteTable("groups", {
  id: text().primaryKey(),
  syncStatus: text().notNull(),
  name: text().notNull(),
  currency: text().notNull(),
  members: text().notNull(),
  createdAt: text().notNull(),
  updatedAt: text().notNull(),
  deletedAt: text(),
});

export const expensesTable = sqliteTable("expenses", {
  id: text().primaryKey(),
  syncStatus: text().notNull(),
  groupId: text().notNull(),
  title: text().notNull(),
  payerName: text().notNull(),
  splitExpense: text().notNull(),
  createdAt: text().notNull(),
  updatedAt: text().notNull(),
  deletedAt: text(),
  receipt: text(),
});

export const syncedTable = (
  name: string,
  columns: Record<string, SQLiteColumnBuilderBase>,
) => {
  return sqliteTable(name, {
    ...columns,
    id: text().primaryKey(),
    syncStatus: text().notNull(),
  });
};

export type SyncableTable = ReturnType<typeof syncedTable>;

export const syncQueue = sqliteTable("sync_queue", {
  id: text().primaryKey(),
  type: text().notNull(),
  table: text().notNull(),
  objectId: text().notNull(),
  data: text(),
  createdAt: text().notNull(),
});

export const tables = {
  groups: groupsTable,
  expenses: expensesTable,
};

export type ExpenseRecord = typeof expensesTable.$inferSelect;
export type GroupRecord = typeof groupsTable.$inferSelect;
export type SyncQueueRecord = typeof syncQueue.$inferSelect;

export type Transaction = SQLiteTransaction<
  "sync",
  SQLiteRunResult | RunResult,
  Record<string, unknown>,
  ExtractTablesWithRelations<Record<string, unknown>>
>;

export type Database = ExpoSQLiteDatabase | BetterSQLite3Database;
