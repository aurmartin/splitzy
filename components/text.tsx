import {
  Text as BaseText,
  StyleSheet,
  type TextProps as BaseTextProps,
} from "react-native";

export type TextProps = BaseTextProps & {
  lightColor?: string;
  darkColor?: string;
  type?:
    | "body"
    | "bodyMedium"
    | "bodySmall"
    | "display"
    | "displayLarge"
    | "headline"
    | "headlineMedium"
    | "headlineSmall"
    | "title"
    | "titleMedium"
    | "titleSmall"
    | "subtitle"
    | "link";
};

export function Text({
  style,
  lightColor,
  darkColor,
  type = "body",
  ...rest
}: TextProps) {
  return (
    <BaseText
      style={[
        type === "body" ? styles.body : undefined,
        type === "bodyMedium" ? styles.bodyMedium : undefined,
        type === "bodySmall" ? styles.bodySmall : undefined,
        type === "display" ? styles.display : undefined,
        type === "displayLarge" ? styles.displayLarge : undefined,
        type === "headline" ? styles.headline : undefined,
        type === "headlineMedium" ? styles.headlineMedium : undefined,
        type === "headlineSmall" ? styles.headlineSmall : undefined,
        type === "title" ? styles.title : undefined,
        type === "titleMedium" ? styles.titleMedium : undefined,
        type === "titleSmall" ? styles.titleSmall : undefined,
        type === "subtitle" ? styles.subtitle : undefined,
        type === "link" ? styles.link : undefined,
        style,
      ].filter(Boolean)}
      {...rest}
    />
  );
}

const styles = StyleSheet.create({
  body: {
    fontSize: 16,
    lineHeight: 24,
    fontWeight: "400",
  },
  bodyMedium: {
    fontSize: 14,
    lineHeight: 20,
    fontWeight: "400",
  },
  bodySmall: {
    fontSize: 12,
    lineHeight: 16,
    fontWeight: "400",
  },
  displayLarge: {
    fontSize: 56,
    fontWeight: "300",
    lineHeight: 64,
  },
  display: {
    fontSize: 32,
    fontWeight: "400",
    lineHeight: 38,
  },
  headline: {
    fontSize: 32,
    fontWeight: "400",
    lineHeight: 40,
  },
  headlineMedium: {
    fontSize: 28,
    fontWeight: "400",
    lineHeight: 36,
  },
  headlineSmall: {
    fontSize: 24,
    fontWeight: "400",
    lineHeight: 32,
  },
  title: {
    fontSize: 22,
    fontWeight: "400",
    lineHeight: 28,
  },
  titleMedium: {
    fontSize: 16,
    fontWeight: "500",
    lineHeight: 24,
  },
  titleSmall: {
    fontSize: 14,
    fontWeight: "500",
    lineHeight: 20,
  },
  subtitle: {
    fontSize: 20,
    fontWeight: "bold",
  },
  link: {
    lineHeight: 30,
    fontSize: 16,
    color: "#0a7ea4",
  },
});
