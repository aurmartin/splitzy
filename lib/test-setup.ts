import { System } from "@/lib/system";
import {
  clearDatabase,
  createDatabase,
  createSupabaseServer,
} from "@/lib/test-utils";

const server = createSupabaseServer();
let system: System;

beforeEach(async () => {
  clearDatabase(system.db);
});

beforeAll(async () => {
  server.listen();
  system = new System(createDatabase());
  await system.initializeMigrations();
});

afterEach(() => {
  server.resetHandlers();
});

afterAll(() => {
  server.close();
  system.dispose();
});

export { server, system };
