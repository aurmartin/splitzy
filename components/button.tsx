import { Pressable, type PressableProps } from "@/components/pressable";
import { Text } from "@/components/text";
import { Colors } from "@/lib/constants";
import { StyleProp, StyleSheet, ViewStyle } from "react-native";

interface ButtonProps extends PressableProps {
  children: React.ReactNode;
  style?: StyleProp<ViewStyle>;
  type?: "primary" | "secondary";
}

export default function Button(props: ButtonProps) {
  const { type = "primary", children, style, ...rest } = props;

  return (
    <Pressable
      {...rest}
      style={[styles.button, styles[type], style]}
      accessibilityRole="button"
    >
      <Text style={styles.text}>{children}</Text>
    </Pressable>
  );
}

const styles = StyleSheet.create({
  primary: {
    backgroundColor: "white",
  },
  secondary: {
    backgroundColor: "transparent",
  },
  button: {
    borderRadius: 40,
    height: 40,
    alignItems: "center",
    justifyContent: "center",
    paddingHorizontal: 16,
  },
  text: {
    color: Colors.primary,
  },
});
