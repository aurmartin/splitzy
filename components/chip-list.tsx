import { primaryHue } from "@/lib/constants";
import React from "react";
import { Pressable, StyleProp, View, ViewStyle } from "react-native";
import { Text } from "./text";

interface ChipProps {
  label: string;
  isSelected: boolean;
  onPress: () => void;
}

function Chip(props: ChipProps) {
  const { label, isSelected, onPress } = props;

  const backgroundColor = React.useMemo(() => {
    if (isSelected) {
      return `hsl(${primaryHue}, 50%, 80%)`;
    }

    return "hsl(0, 0%, 90%)";
  }, [isSelected]);

  const style = {
    padding: 8,
    backgroundColor,
    borderRadius: 16,
  };

  return (
    <Pressable
      onPress={onPress}
      style={style}
      accessibilityLabel={label}
      accessibilityRole="button"
    >
      <Text type="bodyMedium">{label}</Text>
    </Pressable>
  );
}

interface ChipListItem {
  label: string;
  value: string;
}

interface ChipListProps {
  onChange: (value: string) => void;
  value: string;
  items: ChipListItem[];
  style?: StyleProp<ViewStyle>;
}

function ChipList(props: ChipListProps) {
  const { onChange, value, items, style } = props;

  return (
    <View style={[{ flexDirection: "row", gap: 8 }, style]}>
      {items.map((item) => (
        <Chip
          key={item.value}
          label={item.label}
          isSelected={item.value === value}
          onPress={() => onChange(item.value)}
        />
      ))}
    </View>
  );
}

export { ChipList, type ChipListProps };
