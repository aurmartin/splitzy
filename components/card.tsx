import { Pressable } from "@/components/pressable";
import { View } from "react-native";

interface CardProps {
  children: React.ReactNode;
  onPress?: () => void;
}

export default function Card(props: CardProps) {
  if (props.onPress) {
    return (
      <Pressable
        style={{
          borderRadius: 8,
          backgroundColor: "white",
          overflow: "hidden",
          padding: 14,
        }}
        android_ripple={{ color: "hsl(0, 0, 40%)" }}
        onPress={props.onPress}
      >
        {props.children}
      </Pressable>
    );
  } else {
    return (
      <View
        style={{
          backgroundColor: "white",
          borderRadius: 8,
          padding: 14,
        }}
      >
        {props.children}
      </View>
    );
  }
}
