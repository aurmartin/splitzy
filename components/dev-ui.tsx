import { useSystem } from "@/components/system-provider";
import { ExpensesRepository } from "@/lib/expenses/repository";
import { getGroupRow, useAddGroup } from "@/lib/groups";
import { System } from "@/lib/system";
import { generateId } from "@/lib/utils";
import dinero from "dinero.js";
import { usePathname } from "expo-router";
import { useEffect, useState } from "react";
import { Button, Platform, StyleSheet, Text, View } from "react-native";
import { type PercentageSplit } from "@/lib/expenses";

function getRandomMember(members: string[]) {
  return members[Math.floor(Math.random() * members.length)];
}

async function generateExpenses(system: System, groupId: string) {
  const count = 10;
  const group = getGroupRow(system, groupId);
  if (!group) {
    throw new Error("Group not found");
  }
  const members = group.members;

  for (let i = 0; i < count; i++) {
    const payer = getRandomMember(members);

    const total = dinero({
      amount: Math.floor(Math.random() * 100),
      currency: group.currency,
    });

    const splitExpense: PercentageSplit = {
      type: "percentage",
      total,
      members,
      ratios: members.reduce(
        (acc, member) => {
          acc[member] = Math.floor(Math.random() * 100);
          return acc;
        },
        {} as PercentageSplit["ratios"],
      ),
    };

    const id = generateId();
    const date = new Date();
    date.setDate(date.getDate() - i * 3);

    await ExpensesRepository.insertExpense(system, {
      id,
      groupId,
      title: `Expense ${i + 1}`,
      payerName: payer,
      splitExpense,
      createdAt: date,
      updatedAt: date,
      deletedAt: undefined,
      receipt: null,
    });
  }
}

const MemoryUsageInfo = () => {
  const [used, setUsed] = useState(0);

  const formatBytes = (bytes: number) => {
    if (bytes < 1024) {
      return `${bytes} bytes`;
    }
    if (bytes < 1024 * 1024) {
      return `${(bytes / 1024).toFixed(2)} Kb`;
    }
    return `${(bytes / 1024 / 1024).toFixed(2)} Mb`;
  };

  useEffect(() => {
    if (Platform.OS === "web") {
      const interval = setInterval(() => {
        setUsed((performance as any).memory.usedJSHeapSize);
      }, 500);
      return () => clearInterval(interval);
    }
  }, []);

  return <Text style={{ marginBottom: 10 }}>Memory: {formatBytes(used)}</Text>;
};

export default function DevUi() {
  const pathname = usePathname();
  const isGroup = pathname.includes("/groups/");
  const groupId = isGroup ? pathname.split("/groups/")[1] : undefined;
  const system = useSystem();
  const addGroup = useAddGroup();
  const onGenerateExpenses = () => {
    if (!isGroup) {
      return;
    }
    generateExpenses(system, groupId!);
  };

  const onGenerateGroup = () => {
    addGroup({
      name: "Test Group",
      members: ["Alice", "Bob", "Charlie"],
      currency: "EUR",
    });
  };

  return (
    <View style={styles.wrapper}>
      <View style={styles.container}>
        <Text>Dev UI</Text>
        <MemoryUsageInfo />
        <Button
          title="Generate expenses"
          onPress={onGenerateExpenses}
          disabled={!isGroup}
        />
        <Button title="Generate group" onPress={onGenerateGroup} />
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  wrapper: {
    position: "absolute",
    top: 0,
    right: 0,
  },
  container: {
    backgroundColor: "white",
    borderWidth: 1,
    borderColor: "#ccc",
    borderRadius: 4,
    margin: 10,
    padding: 10,
    shadowColor: "black",
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.25,
    shadowRadius: 10,
    gap: 10,
  },
});
