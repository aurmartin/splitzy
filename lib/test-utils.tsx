import { SystemProvider } from "@/components/system-provider";
import { Database, tables } from "@/lib/db/schema";
import { render, type RenderOptions } from "@testing-library/react-native";
import BetterSqliteDatabase from "better-sqlite3";
import { drizzle } from "drizzle-orm/better-sqlite3";
import {
  RenderRouterOptions,
  renderRouter,
  type MockContextConfig,
} from "expo-router/testing-library";
import { HttpResponse, http } from "msw";
import { setupServer } from "msw/node";
import React, { type ReactElement } from "react";
import { System } from "./system";
import * as jws from "jws";

const AllTheProviders = ({
  children,
  system,
}: {
  children: ReactElement;
  system: System;
}) => {
  return <SystemProvider system={system}>{children}</SystemProvider>;
};

const customRender = (
  ui: ReactElement,
  system: System,
  options?: Omit<RenderOptions, "wrapper">,
) => {
  const wrapper = ({ children }: { children: ReactElement }) => (
    <AllTheProviders system={system}>{children}</AllTheProviders>
  );

  return render(ui, { wrapper, ...options });
};

const customRenderRouter = (
  context: MockContextConfig,
  system: System,
  options?: Omit<RenderRouterOptions, "wrapper">,
) => {
  const wrapper = ({ children }: { children: ReactElement }) => (
    <AllTheProviders system={system}>{children}</AllTheProviders>
  );

  return renderRouter(context, { ...options, wrapper });
};

const createDatabase = () => {
  const db = new BetterSqliteDatabase(":memory:");
  return drizzle(db);
};

const clearDatabase = (db: Database) => {
  db.delete(tables.groups).run();
  db.delete(tables.expenses).run();
};

const createSupabaseServer = () =>
  setupServer(
    http.get("http://localhost:50001/rest/v1/groups", () =>
      HttpResponse.json([]),
    ),
    http.get("http://localhost:50001/rest/v1/expenses", () =>
      HttpResponse.json([]),
    ),
    http.post("http://localhost:50001/rest/v1/expenses", () =>
      HttpResponse.json({}),
    ),
    http.patch("http://localhost:50001/rest/v1/expenses", () =>
      HttpResponse.json({}),
    ),
    http.delete("http://localhost:50001/rest/v1/expenses", () =>
      HttpResponse.json({}),
    ),
    http.post("http://localhost:50001/auth/v1/otp", () =>
      HttpResponse.json({}),
    ),
    http.get("http://localhost:50001/auth/v1/user", () =>
      HttpResponse.json({
        id: "test-user-id",
        email: "test@test.com",
        confirmation_sent_at: "2025-03-01T00:00:00Z",
        created_at: "2025-03-01T00:00:00Z",
        updated_at: "2025-03-01T00:00:00Z",
      }),
    ),
  );

const setFakeSession = (system: System) => {
  const access_token = generateAccessToken();
  const refresh_token = "fake-refresh-token";

  system.setSession({ access_token, refresh_token });
};

const generateAccessToken = () => {
  const nowTimestamp = Math.floor(Date.now() / 1000);
  const expiresAt = nowTimestamp + 60 * 60 * 24 * 30; // 30 days

  const payload = {
    exp: expiresAt,
    iat: nowTimestamp,
    email: "test@test.com",
  };

  return jws.sign({
    header: { alg: "HS256" },
    payload,
    secret: "secret",
  });
};

export * from "@testing-library/react-native";
export {
  clearDatabase,
  createDatabase,
  createSupabaseServer,
  customRender as render,
  customRenderRouter as renderRouter,
  setFakeSession,
  generateAccessToken,
};
