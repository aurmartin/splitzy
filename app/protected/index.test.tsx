import { Text } from "@/components/text";
import { groupsTable } from "@/lib/db/schema";
import { system } from "@/lib/test-setup";
import { buildGroupRecord, renderRouter } from "@/lib/test-utils";
import { fireEvent, screen, userEvent } from "@testing-library/react-native";
import GroupsScreen from ".";

const routerContext = {
  "/protected": () => <GroupsScreen />,
  "/protected/groups/:id": () => <Text>Group Details</Text>,
  "/protected/groups/new": () => <Text>New Group</Text>,
};

describe("GroupsScreen", () => {
  it("should render snapshot", () => {
    renderRouter(routerContext, system, { initialUrl: "/protected" });
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it("should render correct title", () => {
    renderRouter(routerContext, system, { initialUrl: "/protected" });

    screen.getByText("Splitzy Dev");
  });

  it("should render groups", async () => {
    const group = buildGroupRecord({
      name: "Group 1",
      members: JSON.stringify(["Alice", "Bob"]),
    });
    await system.db.insert(groupsTable).values(group);

    renderRouter(routerContext, system, { initialUrl: "/protected" });

    screen.getByText("Group 1");
    screen.getByText("Alice, Bob");
  });

  it("should render no members message", async () => {
    const group = buildGroupRecord({ members: JSON.stringify([]) });
    await system.db.insert(groupsTable).values(group);

    renderRouter(routerContext, system, { initialUrl: "/protected" });

    screen.getByText("(no members)");
  });

  it("should render empty state", () => {
    renderRouter(routerContext, system, { initialUrl: "/protected" });
    screen.getByText("No groups found. Create one by using the '+' button!");
  });

  it("should allow the user to refresh the groups", async () => {
    const group = buildGroupRecord({ name: "initial group name" });
    (system.supabaseConnector.selectAll as jest.Mock).mockResolvedValueOnce({
      data: [group],
      error: null,
    });
    await system.syncEngine.syncTableFromRemote(groupsTable);

    renderRouter(routerContext, system, { initialUrl: "/protected" });

    screen.getByText("initial group name");

    const updatedGroup = { ...group, name: "updated group name" };
    (system.supabaseConnector.selectAll as jest.Mock).mockResolvedValueOnce({
      data: [updatedGroup],
      error: null,
    });

    fireEvent(screen.getByTestId("groups-flat-list"), "onRefresh");

    await screen.findByText("updated group name");
  });

  it("should navigate to the new group screen", async () => {
    const user = userEvent.setup();

    renderRouter(routerContext, system, { initialUrl: "/protected" });
    await user.press(screen.getByTestId("fab-button"));

    screen.getByText("New Group");
  });

  it("should navigate to the group details screen", async () => {
    const user = userEvent.setup();
    const group = buildGroupRecord();
    await system.db.insert(groupsTable).values(group);

    renderRouter(routerContext, system, { initialUrl: "/protected" });
    await user.press(screen.getByText(group.name));

    screen.getByText("Group Details");
  });
});
