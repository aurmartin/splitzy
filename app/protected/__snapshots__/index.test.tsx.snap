// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupsScreen should render snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNCSafeAreaView
    edges={
      {
        "bottom": "additive",
        "left": "additive",
        "right": "additive",
        "top": "additive",
      }
    }
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNCSafeAreaView
      edges={
        {
          "bottom": "additive",
          "left": "additive",
          "right": "additive",
          "top": "additive",
        }
      }
      style={
        [
          {
            "backgroundColor": "hsl(0 0% 94%)",
            "flex": 1,
            "padding": 8,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          [
            {
              "fontSize": 32,
              "fontWeight": "400",
              "lineHeight": 38,
            },
            {
              "marginBottom": 16,
              "marginTop": 16,
            },
          ]
        }
      >
        Splitzy Dev
      </Text>
      <RCTScrollView
        ListEmptyComponent={
          <Text
            style={
              {
                "marginVertical": 64,
                "textAlign": "center",
              }
            }
          >
            No groups found. Create one by using the '+' button!
          </Text>
        }
        contentContainerStyle={
          {
            "gap": 8,
          }
        }
        data={[]}
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onRefresh={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        refreshControl={
          <RefreshControlMock
            onRefresh={[Function]}
            refreshing={false}
          />
        }
        refreshing={false}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={0.0001}
        showsVerticalScrollIndicator={false}
        stickyHeaderIndices={[]}
        testID="groups-flat-list"
        viewabilityConfigCallbackPairs={[]}
      >
        <RCTRefreshControl />
        <View>
          <Text
            onLayout={[Function]}
            style={
              [
                {
                  "fontSize": 16,
                  "fontWeight": "400",
                  "lineHeight": 24,
                },
                {
                  "marginVertical": 64,
                  "textAlign": "center",
                },
              ]
            }
          >
            No groups found. Create one by using the '+' button!
          </Text>
        </View>
      </RCTScrollView>
      <View
        style={
          {
            "bottom": 20,
            "position": "absolute",
            "right": 8,
            "zIndex": 100,
          }
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "backgroundColor": "hsl(200 40% 50%)",
                "borderRadius": 16,
                "elevation": 2,
                "padding": 16,
                "shadowColor": "black",
              },
              {
                "overflow": "hidden",
              },
            ]
          }
          testID="fab-button"
        >
          <Text
            allowFontScaling={false}
            selectable={false}
            style={
              [
                {
                  "color": "white",
                  "fontSize": 24,
                },
                undefined,
                {
                  "fontFamily": "ionicons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                {},
              ]
            }
          >
            î¨ˆ
          </Text>
        </View>
      </View>
    </RNCSafeAreaView>
  </RNCSafeAreaView>
</RNCSafeAreaProvider>
`;
